src/common/utils.h

```diff
@@ -364,6 +364,50 @@ inline std::string dev_type_string(const int dev_type) {
   return "unknown";
 }
 
+/*! \brief get string representation of the operator stypes */
+inline std::string operator_stype_string(const nnvm::NodeAttrs& attrs,
+                                         const int dev_mask,
+                                         const std::vector<int>& in_attrs,
+                                         const std::vector<int>& out_attrs) {
+  std::string result = "";
```

Comment:
It's pretty expensive to do a lot of string +='s because it causes lots of reallocations/copies. Maybe better to use stringstream/strstream like you did with the fallback warning?